### PR TITLE
Adopt and implement the full ReQL DSL prerequisites

### PR DESCRIPTION
**Pull request by tbolier**

### CODE REVIEW

File: src/Connection/Connection.php [hunk 1]
```diff
@@ -196,18 +214,34 @@ public function execute(array $query): array
...
+        $this->run($query);
+        $this->noReply = false;
+    }
+
+    /**
+     * @param MessageInterface $query
      * @return mixed
      */
-    private function utf8Converter(array $array): array
+    private function utf8Converter(MessageInterface $query): MessageInterface
```
Comment (marcgeurts) [inline src/Connection/Connection.php]: is MessageInterface $message not better?

File: src/Connection/Connection.php [hunk 2]
```diff
@@ -152,7 +163,7 @@ public function noReplyWait(): void
         // Generate a token for the request
         $token = $this->generateToken();
```
Comment (marcgeurts) [inline src/Connection/Connection.php]: No need for comment, function name is clear enough

File: src/Connection/Connection.php [hunk 3]
```diff
@@ -234,14 +268,22 @@ private function generateToken(): int
...
      * @throws \Exception
      */
-    private function sendQuery($token, $query): int
+    private function sendQuery(int $token, MessageInterface $message): int
     {
-        $request = json_encode($query);
+        $message->setOptions([
+            'db' => [
+                TermType::DB,
+                [$this->dbname],
```
Comment (marcgeurts) [inline src/Connection/Connection.php]: 👍

File: src/Query/Builder.php [hunk 4]
```diff
@@ -3,26 +3,33 @@
...
+    private $rethink;
 
     /**
      * @var TableInterface
      */
     private $table;
 
     /**
-     * @param ManagerInterface $manager
+     * @var MessageInterface
```
[the remark this answers is not in the document]
Comment (tbolier) [reply src/Query/Builder.php]: This is a class property, not a parameter for a method

File: src/Rethink.php [hunk 5]
```diff
@@ -1,48 +1,48 @@
...
+class Rethink implements RethinkInterface
 {
     /**
      * @var ConnectionInterface
      */
     private $connection;
 
     /**
-     * @param ConnectionInterface $connection
+     * @var BuilderInterface
```
[the remark this answers is not in the document]
Comment (tbolier) [reply src/Rethink.php]: This is a class property, not a parameter for a method

**Pull request merged**